Joss paper (#87)
CMakePP/CMakePPLang: paper/paper.md

# CMakePPLang

## Summary

## Statement of Need

> CMake is small, extensible, and exceeds at generating build systems for many combinations of platforms, compilers, and build configurations.

Author: What does "small" mean in this sentence?

CMake has become the *de facto* standard tool for building C, C++, and Fortran
programs of moderate to large size. As the size of a project increases, the
complexity of the CMake code to build it tends to also increase. Object-
oriented programming excels at managing large, complex code bases, and there

> is an increasing need for this in the CMake language.

Author: If you can find a reference (and we have reference slots available) it would be good to cite something here about benefits of OOP.

As more tooling is built around CMake to make building projects with CMake
easier and less error prone, these tools will need to be designed in a

> maintainable and testable way. Object-oriented programming excels at creating

Author: Again a reference would bolster this claim (could quite possibly be the same reference as above).

Reviewer: I can probably check one of my old software engineering books for this reference and the one above.

maintainable code if done correctly. Complexity of builds will also increase as
we move toward heterogeneous systems, requiring programs to leverage a

> combination of CPUs, GPUs, and other accelerator hardware.

Author: ```suggestion
combination of CPUs, GPUs, and other specialized hardware.
```
You can also imagine wanting the build to detect things like the type of network (I'm pretty sure OpenMPI does this in order to optimize the communication), presence of parallel file system, or non-volatile memory.

> Tobias Becker recognized these issues and wrote a purely object-oriented

Author: Reference needed.

language on top of CMake, called CMake++ (formerly oo-cmake).
This project appears to have been abandoned, as there have only been two
commits since July 2017, both in 2021.

CMakePPLang has been developed to provide extensions to the CMake language
which provide objected-oriented functionality and other quality-of-life
improvements. The main features of CMakePPLang are the object-oriented
functionality, addition of a map structure, strong data typing, and 
backwards-compatability with CMake. These features allow for easier general
programming in CMake, which is key to writing complex build tools in the
language.

Currently, CMakePPLang is used within the CMakePP organization as the 

> foundation for two in-progress projects, CMakeTest, which provides a solution

Author: ```suggestion
foundation for two in-progress projects, CMakeTest and CMaize. CMakeTest provides a solution
```

> for unit testing CMake and CMakePPLang code, as well as CMaize, a CMake tool

Author: ```suggestion
for unit testing CMake and CMakePPLang code. CMaize is a CMake tool
```
